File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-3.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Codec() {
    // Encodes a URL to a shortened URL.
    fun serialize(root: TreeNode?): String {
        val data = mutableListOf<String>()

        fun helper(root: TreeNode?) {
            if(root == null) {
                data.add("None")
                return
            }

            data.add(root.`val`.toString())
            helper(root.left)
            helper(root.right)
        }
        helper(root)
        print(data.toString())
        println(data.joinToString(", "))
        return data.joinToString(", ")
    }

    // Decodes your encoded data to tree.
    fun deserialize(data: String): TreeNode? {
        val data = data.split(", ").toMutableList()

        fun helper(root: TreeNode?): TreeNode?
        {
            if(data[0] == "None"){
                data.removeFirst()
                return null
            }

            val root = TreeNode(data.removeFirst().toInt())
            root.left = helper(root)
            root.right = helper(root)

            return root
        }
        return helper(null)
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.ans = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dia(root):
            if not root:
                return 0 

            l = dia(root.left)
            r = dia(root.right)
            
            self.ans = max(self.ans, l +r)

            return 1 + max(l,r) 
        
        dia(root)
        return self.ans